File: src/NStreamCom/Collector.cs
using static NStreamCom.Collector;

namespace NStreamCom
{
    public class Collector
    {
        public enum States
        {
            Collected,
            MissingSize,
            MissingData,
            WaitingSize,
            WaitingData
        }

        public class StateChangedEventArgs(States previousState, States newState) : EventArgs
        {
            public readonly States PreviousState = previousState;

            public readonly States NewState = newState;
        }

        public delegate void StateChangedEventHandler(object? sender, StateChangedEventArgs e);

        public States State { get; private set; } = States.WaitingSize;

        public bool SizeReady { get => State == States.WaitingData; }

        public bool DataReady { get => State == States.Collected; }

        public uint Size
        {
            get
            {
                if (!SizeReady)
                    throw new InvalidOperationException("Size is not ready.");
                return _nextSize;
            }
        }

        public byte[] Data
        {
            get
            {
                if (!DataReady)
                    throw new InvalidOperationException("Data is not ready.");
                return _decoder.Bytes;
            }
        }

        public event StateChangedEventHandler? StateChanged;

        private readonly BufferedDecoder _decoder = new();

        private uint _nextSize = 0;

        private States _previousState = States.WaitingSize;

        public void Reset()
        {
            _nextSize = 0;
            _decoder.Reset();
            State = States.WaitingSize;
        }

        public States Collect(byte b)
        {
            switch (State)
            {
                case States.Collected:
                case States.MissingSize:
                case States.MissingData:
                    Reset();
                    Collect(b);
                    break;
                case States.WaitingSize:
                    CollectSizeByte(b);
                    break;
                case States.WaitingData:
                    CollectDataByte(b);
                    break;
                default:
                    break;
            }

            if (_previousState != State)
            {
                StateChanged?.Invoke(this, new(_previousState, State));
                _previousState = State;
            }

            return State;
        }

        private void CollectSizeByte(byte b)
        {
            if ((b & (1 << 7)) == 0)
            {
                State = States.MissingSize;
                return;
            }

            b &= (byte)~(0xFF << (byte)Sizing.DATA_BITS);
            bool last = _decoder.DecodedIndex == sizeof(uint);
            _decoder.Next(b , last);

            if (!last)
                return;

            _nextSize = BitConverter.ToUInt32(_decoder.Bytes);
            _decoder.Reset();
            State = _nextSize == 0 ? States.Collected : States.WaitingData;
        }

        private void CollectDataByte(byte b)
        {
            if ((b & 0b10000000) > 0)
            {
                State = States.MissingData;
                return;
            }

            bool last = _decoder.DecodedIndex == _nextSize;
            _decoder.Next(b , last);

            if (!last)
                return;

            State = States.Collected;
        }
    }

    public static class StatesExtensions
    {
        public static bool ErrorState(this States state) => state == States.MissingSize || state == States.MissingData;
    }
}

File: src/nstreamcom/collector.py
import enum
from .sizing import DATA_BITS, NSIZE_SIZE
from .buffered_decoder import BufferedDecoder


class CollectorStates(enum.Enum):
    Collected = 0,
    MissingSize = 1,
    MissingData = 2,
    WaitingSize = 3,
    WaitingData = 4


class Collector:
    def __init__(self) -> None:
        self._state: CollectorStates = CollectorStates.WaitingSize
        self._next_size: int = 0
        self._decoder: BufferedDecoder = BufferedDecoder()

    @property
    def state(self) -> CollectorStates:
        return self._state

    @property
    def size_ready(self) -> bool:
        return self._state == CollectorStates.WaitingData or self._state == CollectorStates.Collected

    @property
    def data_ready(self) -> bool:
        return self._state == CollectorStates.Collected

    @property
    def next_size(self) -> int:
        if self.size_ready:
            return self._next_size
        return 0

    @property
    def bytearray(self) -> bytearray:
        if self.data_ready:
            return self._decoder.bytearray
        return bytearray()

    def reset(self) -> None:
        self._next_size = 0
        self._decoder.reset()
        self._state = CollectorStates.WaitingSize

    def collect(self, byte: int) -> CollectorStates:
        match self._state:
            case CollectorStates.Collected:
                self.reset()
                self.collect(byte)
            case CollectorStates.MissingSize:
                self.reset()
                self.collect(byte)
            case CollectorStates.MissingData:
                self.reset()
                self.collect(byte)
            case CollectorStates.WaitingSize:
                self._collect_size_byte(byte)
            case CollectorStates.WaitingData:
                self._collect_data_byte(byte)

        return self._state

    def _collect_size_byte(self, byte: int) -> None:
        if byte & (1 << 7) == 0:
            self._state = CollectorStates.MissingSize
            return

        byte &= ~(0xFF << DATA_BITS)
        last: bool = self._decoder._decoded_index == NSIZE_SIZE
        self._decoder.next(byte , last)

        if not last:
            return

        self._next_size = int.from_bytes(self._decoder.bytearray, 'little')
        self._decoder.reset()
        self._state = CollectorStates.WaitingData if self._next_size else CollectorStates.Collected

    def _collect_data_byte(self, byte: int) -> None:
        if (byte & 1 << 7) > 0:
            self._state = CollectorStates.MissingData
            return

        last: bool = self._decoder.decoded_index == self._next_size
        self._decoder.next(byte, last)

        if not last:
            return

        self._state = CollectorStates.Collected
